### PR TITLE
Chapter 13

### PR DESCRIPTION
**Pull request by Moonforeva**

### CODE REVIEW

**Comment by Moonforeva** [comment]
Ready

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,3 +1,13 @@
 class ApplicationController < ActionController::Base
   include SessionsHelper
+
+  private
+
+  def logged_in_user
+    unless logged_in?
```
Comment (manhbnt-0960) [inline app/controllers/application_controller.rb:7]: let's use guard clause

File: app/controllers/microposts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      flash[:success] = "Micropost created!"
+      redirect_to root_url
+    else
+      @feed_items = current_user.feed.paginate page: params[:page]
+      render :static_pages/home
+    end
+  end
+
+  def destroy
+    @micropost.destroy
```
Comment (manhbnt-0960) [inline app/controllers/microposts_controller.rb:18]: hande case destroy fail

File: app/controllers/static_pages_controller.rb [hunk 3]
```diff
@@ -1,5 +1,10 @@
 class StaticPagesController < ApplicationController
-  def home; end
+  def home 
+    if logged_in?
```
Comment (manhbnt-0960) [inline app/controllers/static_pages_controller.rb:3]: let's use guard clause

File: app/helpers/microposts_helper.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+module MicropostsHelper
```
Comment (manhbnt-0960) [inline app/helpers/microposts_helper.rb:1]: delete this file

File: app/models/micropost.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  has_one_attached :image
+  default_scope -> { order(created_at: :desc) }
```
Comment (manhbnt-0960) [inline app/models/micropost.rb:4]: don't use default scope, let's use normal scope

File: app/models/micropost.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  validates :user_id, presence: true
+  validates :content, presence: true, length: { maximum: 140 }
```
Comment (manhbnt-0960) [inline app/models/micropost.rb:6]: move 140 to settings

**Comment by Moonforeva** [comment]
Ready

File: app/controllers/microposts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+  def create; 
+    @micropost = current_user.microposts.build micropost_params
+    @micropost.image.attach params[:micropost][:image]
+    if @micropost.save
+      flash[:success] = "Micropost created!"
+      redirect_to root_url
+    else
+      @feed_items = current_user.feed.paginate page: params[:page]
+      render :static_pages/home
```
Comment (chungpv-1008) [inline app/controllers/microposts_controller.rb:13]:
```suggestion
      render "static_pages/home"
```

File: app/controllers/microposts_controller.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    end
+  end
+
+  def destroy
+    if @micropost.destroy
+      flash[:success] = "Micropost deleted"
+      redirect_to request.referrer || root_url
+    else
+      flash.now[:danger] = "Micropost not deleted"
```
Comment (chungpv-1008) [inline app/controllers/microposts_controller.rb:22]:
```suggestion
      flash[:danger] = "Micropost not deleted"
```

File: app/controllers/microposts_controller.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+  private
+
+  def micropost_params
+    params.require(:micropost).permit :content, :image
+  end
+
+  def correct_user
+    @micropost = current_user.microposts.find_by id: params[:id]
+    redirect_to root_url if @micropost.nil?
```
Comment (chungpv-1008) [inline app/controllers/microposts_controller.rb:34]:
```suggestion
    redirect_to root_url if @micropost
```

File: app/models/user.rb [hunk 10]
```diff
@@ -66,6 +67,14 @@ def password_reset_expired?
     reset_sent_at < Settings.password_expired_time.hours.ago
   end
   
+  def feed
+    Micropost.where("user_id = ?", id)
```
Comment (chungpv-1008) [inline app/models/user.rb:71]:
```suggestion
    microposts
```

File: app/views/password_resets/edit.html.erb [hunk 11]
```diff
@@ -3,8 +3,8 @@
 
 <div class="row">
   <div class="col-md-6 col-md-offset-3">
-    <%= form_for @user, url: password_reset_path(params[:id]) do |f| %>
-      <%= render "shared/error_messages" %>
+    <%= form_for @user, url: password_reset_path params[:id] do |f| %>
```
Comment (chungpv-1008) [inline app/views/password_resets/edit.html.erb:6]:
```suggestion
    <%= form_for @user, url: password_reset_path(params[:id]) do |f| %>
```

File: app/views/shared/_error_messages.html.erb [hunk 12]
```diff
@@ -1,10 +1,10 @@
-<% if @user.errors.any? %>
+<% if object.errors.any? %>
   <div id="error_explanation">
     <div class="alert alert-danger">
-      The form contains <%= pluralize @user.errors.size, "error" %>
+      The form contains <%= pluralize(object.errors.count, "error") %>.
```
Comment (chungpv-1008) [inline app/views/shared/_error_messages.html.erb:4]:
```suggestion
      The form contains <%= pluralize(object.errors.size, "error") %>.
```

File: app/views/shared/_feed.html.erb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+<% if @feed_items.any? %>
+  <ol class="microposts">
+    <%= render @feed_items %>
+  </ol>
+  <%= will_paginate @feed_items, params: { controller: :static_pages, action: :home } %>
```
Comment (chungpv-1008) [inline app/views/shared/_feed.html.erb:5]:
```suggestion
  <%= paginate @feed_items, params: { controller: :static_pages, action: :home } %>
```

File: app/views/shared/_user_info.html.erb [hunk 14]
```diff
@@ -0,0 +1,4 @@
+<%= link_to gravatar_for(current_user, size: 50), current_user %>
+<h1><%= current_user.name %></h1>
+<span><%= link_to "view my profile", current_user %></span>
+<span><%= pluralize(current_user.microposts.count, "micropost") %></span>
```
Comment (chungpv-1008) [inline app/views/shared/_user_info.html.erb:4]:
```suggestion
<span><%= pluralize(current_user.microposts.size, "micropost") %></span>
```

File: app/views/users/show.html.erb [hunk 15]
```diff
@@ -8,5 +8,14 @@
       <%= @user.name %>
       </h1>
     </section>
-  </aside>  
+  </aside> 
+  <div class="col-md-8">
+    <% if @user.microposts.any? %>
+      <h3>Microposts (<%= @user.microposts.count %>)</h3>
```
Comment (chungpv-1008) [inline app/views/users/show.html.erb:14]:
```suggestion
      <h3>Microposts (<%= @user.microposts.size %>)</h3>
```

File: app/views/users/show.html.erb [hunk 16]
```diff
@@ -8,5 +8,14 @@
...
+      <ol class="microposts">
+        <%= render @microposts %>
+      </ol>
+      <%= will_paginate @microposts %>
```
Comment (chungpv-1008) [inline app/views/users/show.html.erb:18]:
```suggestion
      <%= paginate @microposts %>
```

File: config/routes.rb [hunk 17]
```diff
@@ -12,4 +12,5 @@
   resources :users
   resources :account_activations, only: :edit
   resources :password_resets
+  resources :microposts, only: [:create, :destroy]
```
Comment (chungpv-1008) [inline config/routes.rb:15]:
```suggestion
  resources :microposts, only: %i(create destroy)
```

**Comment by Moonforeva** [comment]
Ready

File: app/controllers/microposts_controller.rb [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+    end
+  end
+
+  def destroy
+    if @micropost.destroy
+      flash[:success] = "Micropost deleted"
+      redirect_to request.referrer || root_url
+    else
+      flash[:danger] = "Micropost not deleted"
+      redirect_to root_url
```
Comment (saobv-2223) [inline app/controllers/microposts_controller.rb:23]:
```suggestion
    if @micropost.destroy
      flash[:success] = "Micropost deleted"
      redirect_to request.referrer || root_url
    else
      flash[:danger] = "Micropost not deleted"
      redirect_to root_url
    end
```

File: app/controllers/static_pages_controller.rb [hunk 19]
```diff
@@ -1,5 +1,10 @@
 class StaticPagesController < ApplicationController
-  def home; end
+  def home 
+    return if logged_in?
+    
+    @micropost = current_user.microposts.build
+    @feed_items = current_user.feed.paginate(page: params[:page])
```
Comment (saobv-2223) [inline app/controllers/static_pages_controller.rb:6]:
```suggestion
    @feed_items = current_user.feed.paginate(page: params[:page]).per `number`
```

File: app/models/micropost.rb [hunk 20]
```diff
@@ -0,0 +1,11 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  has_one_attached :image
+  scope -> { order(created_at: :desc) }
```
Comment (saobv-2223) [inline app/models/micropost.rb:4]:
```suggestion
  scope -> { order created_at: :desc }
```

File: app/views/shared/_micropost_form.html.erb [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+<% end %>
+
+<script type="text/javascript">
+  $("#micropost_image").bind("change", function() {
+    var size_in_megabytes = this.files[0].size/1024/1024;
+    if (size_in_megabytes > 5) {
+      alert("Maximum file size is 5MB. Please choose a smaller file.");
+    }
+  });
+</script>
```
Comment (saobv-2223) [inline app/views/shared/_micropost_form.html.erb:19]: You should write js code to another js file and call it here

**Comment by Moonforeva** [comment]
Ready

File: app/controllers/microposts_controller.rb [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+
+  private
+
+  def micropost_params
+    params.require(:micropost).permit :content, :image
+  end
+
+  def correct_user
+    @micropost = current_user.microposts.find_by id: params[:id]
+    redirect_to root_url if @micropost
```
Comment (chungpv-1008) [inline app/controllers/microposts_controller.rb:35]:
```suggestion
    redirect_to root_url unless @micropost
```

**Review by chungpv-1008 — COMMENTED** [review]
Please fix comment, then merge this p/r yourself

**Pull request merged**